docs: add guide for contributing new clients (#5628)
confluentinc/ksql: docs/developer-guide/ksqldb-clients/contributing.md

---
layout: page
title: Contributing a new ksqlDB client
tagline: Contributing a new ksqlDB client
description: Contribute a new ksqlDB client for your favorite language.
keywords: ksqlDB, client
---

> This page contains information and recommendations for how to contribute a new [ksqlDB client](index.md).

Reviewer: Maybe we could say something more inviting, like: "If you're interested in using ksqlDB from a programming language we don't yet support, we've built out a guide to make it easy to add your own client. .... "

Functionality
-------------

> Clients should support the following functionality:

Reviewer 2: Should or must?

Author: Good question. Depends on the resolution to the discussion in https://github.com/confluentinc/ksql/pull/5628/files#r441087931.

Author: Based on discussion below, I think we should leave this as "should".

- [Push and pull queries](#push-and-pull-queries)
- [Terminating push queries](#terminating-push-queries)
- [Inserting new rows of data into existing ksqlDB streams](#inserting-new-rows-of-data-into-existing-ksqldb-streams)
- [Creation of new streams, tables, and persistent queries](#creation-of-new-streams-tables-and-persistent-queries)
- [Terminating persistent queries](#terminating-persistent-queries)
- [Listing existing streams, tables, topics, and queries](#listing-existing-streams-tables-topics-and-queries)

### Push and pull queries ###

> Push and pull queries may be issued by sending requests to the [`/query-stream` server endpoint](TODO).

Author: Waiting on docs for the new server endpoints for these links.

Reviewer: Are these HTTP requests? Im confused about how to interact with the server right off the bat. We probably need an opening section about how to generally communicate between the client+server.

> If no errors are encountered while processing the query statement, the server will write a

Reviewer: I can parse what kind of response I should expect from the server, but I don't understand what I need to supply in the request. It might be good to divide these categories up into subsections and make tables out of them. This is the kind of information that's really important to be able to correctly reference as you work.

Reviewer: Ditto with the rest of the sections. Im not sure what I should really expect to supply in the request or expect in the response. If all of that will show in the server docs, we should note that every time.

Author: Good call. I've explicitly called out that implementers should reference the server docs in each section, and added information in places where more details were needed.

metadata row to the response, followed by one or more result rows.

The metadata includes:
- result column names
- result column types, e.g., `STRING`, `INTEGER`, `DOUBLE`, `DECIMAL(4, 2)`, `ARRAY<STRING>`, ``STRUCT<`F1` STRING, `F2` INTEGER>``, etc.
- query ID, in the case of push queries

> These metadata fields should be exposed from the client API. In particular, the query ID is used for

Reviewer 2: ```suggestion
The client API must expose these metadata fields from the client API. In particular, the query ID is used for
```

[terminating push queries](#terminating-push-queries).

> The client representation of result rows should support all of the different data types supported by ksqlDB:

Reviewer 2: Are any optional?

Author: Nope, updated "should" to "must".

| Data type | Example column type string in metadata |
|-----------|----------------------------------------|
| STRING    | STRING                                 |
| INTEGER   | INTEGER                                |
| BIGINT    | BIGINT                                 |
| DOUBLE    | DOUBLE                                 |
| BOOLEAN   | BOOLEAN                                |
| DECIMAL   | DECIMAL(4, 2)                          |
| ARRAY     | ARRAY<INTEGER>                         |
| MAP       | MAP<STRING, DOUBLE>                    |
| STRUCT    | STRUCT<\`F1\` STRING, \`F2\` INTEGER>  |

Array, map, and struct types may be recursively nested within each other, and may contain any of the other types as well.

Users of the client may wish to receive result rows in different ways:
- Do not block. Rather, perform an action asynchronously on each new row as it arrives
- Block until the next row arrives
- Block until all rows have arrived. This only makes sense in the context of pull queries and push queries with `LIMIT` clauses. 

> Details of client interfaces to support these methods will likely vary by language. For example, the [Java client](java-client.md)

Reviewer 2: ```suggestion
Details of client interfaces to support these methods vary by language. For example, the [Java client](java-client.md)
```

uses [Reactive Streams](http://www.reactive-streams.org/) for its asynchronous streaming interface. 

### Terminating push queries ###

> Requests to terminate push queries should be sent to the [`/close-query` server endpoint](TODO).

Reviewer 2: ```suggestion
Terminate push queries by sending requests to the [`/close-query` server endpoint](TODO).
```

> The client should expose an interface that allows users to pass in a push query ID to be passed to this endpoint.

Reviewer 2: ```suggestion
The client must expose an interface that enables users to pass in a push query ID to be passed to this endpoint.
```

### Inserting new rows of data into existing ksqlDB streams ###

ksqlDB supports inserting new rows of data into existing ksqlDB streams via the [`/inserts-stream` server endpoint](TODO).
If the server successfully validates the target stream name from a request made to the `/inserts-stream` endpoint,

> the client may then write JSON-serialized rows of data to the connection, and the server will insert each

Reviewer 2: ```suggestion
the client may then write JSON-serialized rows of data to the connection, and the server inserts each
```

> asynchronously and write an ack containing the relevant sequence number to the response.

Reviewer 2: ```suggestion
asynchronously and writes an ack containing the relevant sequence number to the response.
```

Because the insertions happen asynchronously on the server side, it makes sense for the client to expose APIs
for inserting one row at a time, but inserting batches of rows is less useful due to the lack of transactional guarantees.

The client should provide a convenient way for users to create rows of data to insert.
The Java client's [KsqlObject](TODO) is one example, though details will likely vary by language.

### Creation of new streams, tables, and persistent queries ###

> These DDL/DML requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).

Reviewer 2: ```suggestion
Send these DDL/DML requests to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
```

### Terminating persistent queries ###

> These requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).

Reviewer 2: ```suggestion
Send requests to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
```

### Listing existing streams, tables, topics, and queries ###

> These admin operation requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).

Reviewer 2: ```suggestion
Send these admin operation requests to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
```

> Configurations

Reviewer: This part could use a little more substance. Maybe linking to an example in the Java client would help? I read this and Im not quite sure what to do as an implementer.

Author: Linked to the Java client API docs.

--------------

> The client of course needs to expose options for specifying the address of the ksqlDB server to connect to.

Reviewer 2: ```suggestion
The client must expose options for specifying the address of the ksqlDB server to connect to.
```

> Additional options important to expose include: 

Reviewer 2: ```suggestion
Additional options that are important to expose include: 
```

- Support for TLS-enabled ksqlDB servers
- Support for mutual-TLS-enabled ksqlDB servers
- Support for ksqlDB servers with HTTP basic authentication enabled

It may also be nice to support configuring multiple ksqlDB server addresses, so the client may route requests
to a different server if one is down.

Process
-------

> Thanks for your interest in contributing a client!

Reviewer: Might want to move this to the top.

Author: Wasn't sure how to move the entire sentence to the top without clashing with the current intro, but I did add a bit of thanks to the top as well. LMK if you had something else in mind.

Reviewer 3: I don't think we should assume anyone writing a client will contribute it back. I'd have the 'how to write a client' section at the top (as it currently is), and then have a 'how to contribute your client to the project' section below.  We should support people just wanting to link their public client repo, as well as those wanting to submit the code to our repo.

The latter has challenges around how we compile such a multi-language project.  The Kafka client's team are probably good people to talk to about this. @edenhill or @ryanatconfluent may be good people to talk to about this.

Author: That's a really good point, I hadn't thought about the logistics of compiling a multi-language project. Does it make sense to have clients from other languages in the main ksql repository, or should they be separate repositories?

To get started:

> - Open [KLIP](../../../design-proposals/README.md) to propose the client you'd like to implement. The KLIP should include a high-level design and example interfaces.

Author: Is the expectation that if someone wants to tackle a client, they're expected to complete the project (implement all the functionality listed above)? Seems like quite the commitment. Might be good to add a sentence clarifying the expectation one way or the other so readers aren't left wondering.

Reviewer 3: I would say a client can choose to implement some or all of the functionality.  This will become more important as new features are added to the client API.

This allows contributes to submit clients that just support a minimal subset of operations, rather than putting them off by saying 'it's all or nothing mate'. Such minimal clients can be extended by others as required.

Of course, we can say that anyone supporting push queries _must_ support terminating them as well.

Side question: if the client doesn't terminate the query, but just disconnects, are the queries terminated?

Author: > Side question: if the client doesn't terminate the query, but just disconnects, are the queries terminated?

Push queries are terminated when the underlying connection is closed, or by the server after 10 minutes of unresponsiveness from the client.

- Contribute code to [ksql repo](https://github.com/confluentinc/ksql)

> - Testing: Besides unit tests in the relevant language, there should also be integration tests to spin up a ksqlDB server and validate client behavior. TODO how to make this easier?

Author: Hey @MichaelDrogalis we'll have to think about how to make the addition of integration tests in other languages easier. It's not clear to me how to do this right now. Is there an established way to run code from multiple languages (the ksqlDB server in Java + the client in a non-Java language) for integration tests?

Reviewer: I'm really not sure. Might be a good thing to bring up at the next team meeting, I'm sure we can brainstorm ideas.

---
layout: page
title: Contributing a new ksqlDB client
tagline: Contributing a new ksqlDB client
description: Contribute a new ksqlDB client for your favorite language.
keywords: ksqlDB, client
---

If you're interested in using ksqlDB from a programming language we don't yet support, we've created

> this guide to make it easy to add your own [ksqlDB client](index.md). Thanks for your contribution!

Reviewer 3: Could do with a preamble about the general design of the protocol e.g. 

> Clients communicate with the ksqlDB server via HTTP. Admin requests are submitted using POST requests. Query results and inserts are streaming over HTTP2.  The sections below cover the different types of functionality a client must/could (TBD) support and links to the appropriate API documentation., where example request and responses can be found.

Not really sure if that's correct or not, but you hopefully get the idea: we need to explain to the would be client engineer the general architecture and what to expect in this pge.

metadata row to the response, followed by one or more result rows.

The metadata includes:
- result column names
- result column types, e.g., `STRING`, `INTEGER`, `DOUBLE`, `DECIMAL(4, 2)`, `ARRAY<STRING>`, ``STRUCT<`F1` STRING, `F2` INTEGER>``, etc.
- query ID, in the case of push queries

The client must expose these metadata fields as part of its API. In particular, the query ID is used for
[terminating push queries](#terminating-push-queries).

The client representation of result rows must support all of the different data types supported by ksqlDB:

| Data type | Example column type string in metadata |
|-----------|----------------------------------------|
| STRING    | STRING                                 |
| INTEGER   | INTEGER                                |
| BIGINT    | BIGINT                                 |
| DOUBLE    | DOUBLE                                 |
| BOOLEAN   | BOOLEAN                                |
| DECIMAL   | DECIMAL(4, 2)                          |
| ARRAY     | ARRAY<INTEGER>                         |
| MAP       | MAP<STRING, DOUBLE>                    |

> | STRUCT    | STRUCT<\`F1\` STRING, \`F2\` INTEGER>  |

Author: I've added the link but also the left the table (clarifying that they are merely examples) as I think the table is helpful for showing exactly how the different types are represented by the server endpoint. I considered moving the table into the server endpoint docs but the level of detail seemed a bit overkill, especially compared to the rest of the server endpoint docs.

If the server response for a SQL statement submitted to the `/ksql` endpoint has HTTP status code 200
but the response body does not fit the form documented above, then the submitted SQL statement did
not correspond to a DDL/DML statement. The client should handle this potential scenario by returning
an error to the user.

### Terminating persistent queries ###

Send these requests to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).

The SQL statement that should be sent in the HTTP request body is `TERMINATE <persistent_query_id>;`.

As noted in the [server endpoint documentation](../ksqldb-rest-api/ksql-endpoint.md),
the format of the response body for terminate query requests is the same as for other DDL/DML requests
[above](#creation-and-deletion-of-streams-and-tables).

Configurations
--------------

The client must expose options for specifying the address of the ksqlDB server to connect to.
Additional options that are important to expose include:
- Support for TLS-enabled ksqlDB servers
- Support for mutual-TLS-enabled ksqlDB servers

> - Support for ksqlDB servers with HTTP basic authentication enabled

Reviewer 3: support for ksqlDB servers with custom auth?

Author: What kind of custom auth are you thinking of? If we're talking about situations where clients need to pass custom HTTP auth headers, those aren't currently supported by the Java client. The way to support them would be to add an option to specify custom headers to be added to all requests.

Reviewer 3: The server supports a plugging in custom authentication mechanisms.  This means someone could write a client that was compatible. Not sure what that means in the context of this document. Maybe its sufficient to just call out this is possible.

Author: Ack. Clarified along with the wording change suggested below (regarding supporting multiple ksqlDB server addresses).

As an example, users specify options for the Java client via a [ClientOptions](api/io/confluent/ksql/api/client/ClientOptions.html)
object that is passed when [creating a Client instance](api/io/confluent/ksql/api/client/Client.html#create(io.confluent.ksql.api.client.ClientOptions)).

> It may also be nice to support configuring multiple ksqlDB server addresses, so the client may route requests

Reviewer 3: ```suggestion
It is encouraged that clients support multiple ksqlDB server addresses, so the client may route requests
```

Not that our CLI does! (Unless that's changed recently)

Author: Does our CLI actually support this? I know the REST client (and therefore the CLI) supports parsing multiple server addresses, but AFAIA it only ever uses the first one (i.e., no fallback logic if the first one becomes unavailable): https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java#L110

Reviewer 3: No... you obviously missed the last line of my comment.

Of course, the CLI _should_ support this ;)

Author: Oh LOL I read your comment as "Note that our CLI does" rather than "Not that our CLI does" -- I read it two or three times the same way each time :P Glad we're on the same page.
